File: python/2022/day5/main.py
data = open("python/2022/day5/test.txt").read().split("\n\n")
part1 = [[y.strip("[] ") for y in x if not y.isspace()]for x in zip(*[[line[i : i + 4] for i in range(0, len(line), 4)]for line in [x for x in data[0].split("\n")][:-1]])]
part2 = part1.copy()

for op in [[int(i) for i in x.split() if i.isdigit()] for x in data[1].split("\n")]:
    part1[op[2] - 1] = list(reversed(part1[op[1] - 1][: op[0]])) + part1[op[2] - 1]
    part1[op[1] - 1] = part1[op[1] - 1][op[0] :]
    part2[op[2] - 1] = part2[op[1] - 1][: op[0]] + part2[op[2] - 1]
    part2[op[1] - 1] = part2[op[1] - 1][op[0] :]

print("".join([x[0] for x in part1 if x]), "".join([x[0] for x in part2 if x]), sep="\n")

File: typescript/2022/day5/main.ts
import * as fs from "fs";

const data = fs
  .readFileSync("typescript/2022/day5/test.txt", "utf-8")
  .split("\r\n\r\n");

const res = data[0]
  .split("\r\n")
  .slice(0, data[0].split("\r\n").length - 1)
  .map((item: string) => {
    return item.match(/.{1,4}/g);
  });

const zip = res[0]
  .map((_: string, index: number) => res.map((row: string[]) => row[index]))
  .map((col: string[]) =>
    col
      .map((item) => {
        return item.trim().length !== 0 && _.trim(item, "[] ");
      })
      .filter((i: any) => i !== false)
  );

data[1]
  .split("\r\n")
  .map((move: string) =>
    move
      .split("")
      .map((char: string) => {
        return !isNaN(parseInt(char, 10)) && Number(char);
      })
      .filter((i: any) => i !== false)
  )
  .map((move: number[]) => {
    zip[move[2] - 1] = [
      ...zip[move[1] - 1].slice(0, move[0]).reverse(),
      ...zip[move[2] - 1],
    ];
    zip[move[1] - 1] = zip[move[1] - 1].slice(move[0], zip[move[1] - 1].length);
  });

const ans = zip.map((col: string[]) => col[0]).join("");
ans;
